src/signal-handling.c

```diff
@@ -184,7 +185,93 @@ JL_DLLEXPORT int jl_profile_is_buffer_full(void)
     // Declare buffer full if there isn't enough room to sample even just the
     // thread metadata and one max-sized frame. The `+ 6` is for the two block
     // terminator `0`'s plus the 4 metadata entries.
-    return bt_size_cur + ((JL_BT_MAX_ENTRY_SIZE + 1) + 6) > bt_size_max;
+    return profile_bt_size_cur + ((JL_BT_MAX_ENTRY_SIZE + 1) + 6) > profile_bt_size_max;
+}
+
+void jl_profile_task(void)
+{
+    if (jl_profile_is_buffer_full()) {
+        // Buffer full: Delete the timer
+        jl_profile_stop_timer();
+        return;
+    }
+
+    arraylist_t *tasks = jl_get_all_tasks_arraylist();
+    jl_task_t *t = NULL;
+    uint64_t seed = jl_rand();
+
+    const int n_max_random_attempts = (1 << 16);
+    // randomly select a task that is not done
+    for (int i = 0; i < n_max_random_attempts; i++) {
+        t = (jl_task_t*)tasks->items[cong(tasks->len, &seed)];
+        assert(t == NULL || jl_is_task(t));
+        if (t == NULL) {
+            continue;
+        }
+        int t_state = jl_atomic_load_relaxed(&t->_state);
+        if (t_state == JL_TASK_STATE_DONE) {
+            continue;
+        }
+        break;
+    }
+    // sequentially select a task that is not done
+    for (int i = 0; i < tasks->len; i++) {
+        t = (jl_task_t*)tasks->items[i];
+        assert(t == NULL || jl_is_task(t));
+        if (t == NULL) {
+            continue;
+        }
+        int t_state = jl_atomic_load_relaxed(&t->_state);
+        if (t_state == JL_TASK_STATE_DONE) {
+            continue;
+        }
+        break;
+    }
+    assert(t != NULL);
+
+    arraylist_free(tasks);
+    free(tasks);
+
+    jl_bt_element_t *bt_data_prof = (jl_bt_element_t*)(profile_bt_data_prof + profile_bt_size_cur);
+    size_t bt_size_max = profile_bt_size_max - profile_bt_size_cur - 1;
+
+    jl_record_backtrace_result_t r = jl_record_backtrace(t, bt_data_prof, bt_size_max, 1);
+
+    int tid = r.tid;
+    // we failed to get a backtrace
+    if (r.bt_size == 0) {
+        return;
+    }
```

Comment:
is it okay to just not record anything if we fail to get a backtrace? That changes the effective sample rate, which i think is going to affect downstream analysis of the results.

Also, do we have any data on how frequently this occurs / what it depends on?

Can the existing `@profile` ever fail to get a backtrace?

I think one option (maybe @NHDaly suggested this) is to consider adding some kind of "fake" backtrace that indicates "sample failed", so at least a failure to sample is present in the data.
I think another option (i think maybe the Go profilers do this?) is to somehow return to the user what the true sample rate was. 

Might be worth asking Todd's perspective on this.

Reply:
+1 
I think putting an indicator in here that the sample failed is a good idea. And ideally, also _why_ the sample failed: the task finished before sampling it, the task got scheduled right before we sampled it(?), are there other reasons?

Then I think the profile visualization tools could decide whether to display them or skip them and indicate a variable sample rate (if they support that).

Reply:
also, i guess the other option is to just try again? 🤔  Maybe up to like 5 retries?

I don't know if that would have any impact on the stats, but naively it doesn't seem like it to me

Reply:
We now have C functions `failed_to_stop_thread_fun` and `failed_to_sample_task_fun` which will be used as place-holders in the backtraces whenever we fail to stop a thread or sample a task.